# Tensor 생성
- 파이토치에서 데이터를 저장하는 자료구조
- ndarray와 성격, 사용법이 유사하다.

> **pytorch의 tensor는 숫자 데이터 타입만 지원한다.**

##  원하는 형태(shape) 텐서 생성
- **torch.tensor(자료구조 \[, dtype\])**
    - 지정한 dtype(Data type)에 맞는 Tensor객체를 생성해서 반환한다.
      
## 특정 타입의 Tensor를 직접 생성
- torch.tensor()로 생성하면서 dtype을 지정하면 아래 타입의 Tensor객체가 생성된다.
- 원하는 Type의 Tensor클래스를 이용해 직접 생성해도 된다.
- **torch.FloatTensor(자료구조)**
    - float32 타입 텐서 생성
- **torch.LongTensor(자료구조)**
    - int64 타입 텐서생성
- 그외
    - BoolTensor(bool), CharTensor(int8), ShortTensor(int16), IntTensor(int32), DoubleTensor(float64)
    
## tensor 상태 조회
- **tensor.shape, tensor.size(\[축번호\])**
    -  tensor의 shape조회
- **tensor.dtype, tensor.type()**
    - tensor 원소들의 데이터타입 조회
    - dtype은 **data type**을 type()은 tensor **객체의 클래스 타입**을 반환한다.
- **tensor.ndim, tensor.dim()**  : tensor 차원
- **tensor.numel()**: 전체 원소 개수



In [1]:
import torch
import numpy as np

In [8]:
np.array([1, 2, 3], dtype=np.int16)#"int8")

array([1, 2, 3], dtype=int16)

In [13]:
print('torch Data types')
print("float", torch.float16, torch.float32, torch.float64, torch.float, torch.double)  #torch.float: float32
print("int:", torch.int8, torch.int16, torch.int32, torch.int64, torch.short, torch.int, torch.long)# 가능
print("uint:", torch.uint8, torch.uint16, torch.uint32, torch.uint64)
print("bool:", torch.bool)

torch Data types
float torch.float16 torch.float32 torch.float64 torch.float32 torch.float64
int: torch.int8 torch.int16 torch.int32 torch.int64 torch.int16 torch.int32 torch.int64
uint: torch.uint8 torch.uint16 torch.uint32 torch.uint64
bool: torch.bool


In [10]:
a = torch.tensor([[1,2],[3,4]])#, dtype=torch.float32)


print("shape:", a.shape, a.size())
print("0축 크기:", a.shape[0], a.size(0))
print("type:", a.type(), a.dtype)  # type(): Tensor 객체 타입. dtype: data type => 둘은 좀 다르다.
# type: torch.LongTensor torch.int64

print('차원크기:', a.dim(), a.ndim)
print('원소개수:', a.numel())

print("device:", a.device)  # tensor를 다루는 processor(메모리위치) - cpu, cuda(gpu), mps

shape: torch.Size([2, 2]) torch.Size([2, 2])
0축 크기: 2 2
type: torch.LongTensor torch.int64
차원크기: 2 2
원소개수: 4
device: cpu


In [ ]:
aa = torch.tensor(range(10))
print(aa)
aa.type()

tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])


'torch.LongTensor'

In [17]:
#Float, Double(32, 64bit 실수)/Int, Long(32, 64 bit 정수) type Tensor
b = torch.FloatTensor([1,3,7])  #float32
print(b.dtype)
c = torch.IntTensor([10,20,30])    # int32
print(c.dtype)
d = torch.DoubleTensor([1, 2, 3])  # float64
print(d.dtype)
e = torch.LongTensor([10, 20, 30]) # int64
print(e.dtype)

torch.float32
torch.int32
torch.float64
torch.int64


## 동일한 값을 원소로 가지는 Tensor 생성
- **torch.zeros(\*size), zeros_like(텐서)**: 0으로 구성된 tensor 생성
- **torch.ones(\*size), ones_like(텐서)**: 1로 구성된 tensor생성
- **torch.full(size, fill_value), full_like(텐서, fill_value)**: 지정한 값으로 구성된 tensor생성
  - size
    - shape을 지정한다.
    - zeros, ones는 가변인자이므로 축별 size를 순서대로 서정
    - full 은 tuple/list로 shape을 지정한다.
    

In [20]:
a = torch.zeros(3,2,3) # 3 x 2 x 3
a.dtype
torch.ones(2,3)  # 2 X 3
torch.full([3,2], fill_value=100) # 3 x 2


tensor([[100, 100],
        [100, 100],
        [100, 100]])

In [23]:
# 특정 배열과 동일한 shape의 tensor 생성.
a = torch.tensor([[1, 2],[3, 4]]) # 2 x 2

print(a.shape)
b = torch.zeros_like(a)  # a와 같은 shape, dtype 으로 생성.
b = torch.ones_like(a)
b = torch.full_like(a, 20)
print(b.shape, b.dtype)
b

torch.Size([2, 2])
torch.Size([2, 2]) torch.int64


tensor([[20, 20],
        [20, 20]])

## 동일한 간격으로 떨어진 값들로 구성된 배열생성
- **torch.arange(start=0, end, step=1)**
- **torch.linspace(start, end, steps,)** : steps - 원소개수

In [26]:
torch.arange(10)  # end. 0 ~ 10-1 1씩 증가
torch.arange(0, 1, 0.1) # 0 ~ 1-0.1, 0.1
torch.arange(10, 1, -1) # 10 ~ 1-(-1), -1

tensor([10,  9,  8,  7,  6,  5,  4,  3,  2])

In [28]:
torch.linspace(0, 10, 5) # 0 ~ 10, 5개원소
torch.linspace(0, 1, 11)


tensor([0.0000, 0.1000, 0.2000, 0.3000, 0.4000, 0.5000, 0.6000, 0.7000, 0.8000,
        0.9000, 1.0000])

## 빈 tensor 생성
- **torch.empty(\*size)**

In [29]:
torch.empty(3,2,7) # 3 X 2 X 7

tensor([[[0.0000e+00, 0.0000e+00, 3.6013e-43, 0.0000e+00, 1.9246e-10,
          4.5324e-41, 9.1084e-44],
         [0.0000e+00, 2.4643e-22, 0.0000e+00, 2.4642e-22, 0.0000e+00,
          4.4842e-44, 0.0000e+00]],

        [[4.4842e-44, 0.0000e+00, 1.5831e-40, 0.0000e+00, 0.0000e+00,
          1.4013e-45, 1.7937e-43],
         [0.0000e+00, 1.1210e-43, 0.0000e+00, 2.5010e-22, 0.0000e+00,
          0.0000e+00, 0.0000e+00]],

        [[0.0000e+00, 0.0000e+00, 4.6243e-44, 0.0000e+00, 2.4909e-22,
          0.0000e+00, 2.4827e-22],
         [0.0000e+00, 5.8294e-43, 0.0000e+00, 1.5695e-43, 0.0000e+00,
          2.4711e-22, 0.0000e+00]]])

## 난수를 이용한 생성

- **torch.rand(\*size)**: 0 ~ 1사이 실수로 구성된 배열을 생성. 각 값은 균등분포를 따른다.
- **torch.randn(\*size)**: 표준정규분포(평균:0, 표준편차:1)를 따르는 실수로 구성된 배열 생성
- **torch.randint(low=0, high, size)**: 지정한 범위의 정수로 구성된 배열 생성
- **torch.randperm(n)**: 0 ~ n-1 사이의 정수를 랜덤하게 섞은 값을 원소로 가지는 배열 생성

In [46]:
torch.manual_seed(0)  # seed 설정
torch.rand(1, 3, 5) # 100 x 3 x 5
torch.randn(30, 3) # 30 X 3
torch.randint(1, 100, (3, 3, 6))
torch.randperm(100) # 0 ~ 99 를 섞어서 구성

tensor([18, 41, 22, 48, 26,  3, 42, 40, 46, 14, 90, 59, 82, 10, 36, 45,  0, 65,
        33, 63, 64, 85, 89, 49, 60, 44, 52, 50, 98,  1, 93, 78, 56, 77,  7, 34,
        67, 29,  5, 86, 20, 91, 15, 97, 73, 68, 31, 80, 55, 92, 53, 37, 58, 47,
        51, 54, 75, 61, 74, 16, 25,  9, 12,  2,  6, 27, 30,  4, 19, 87, 32, 39,
        23, 17, 96, 21, 79, 84, 24, 72, 83, 69, 70, 99, 76, 43, 35,  8, 88, 13,
        62, 95, 28, 94, 38, 11, 71, 81, 66, 57])

In [49]:
a = torch.arange(100)
a
# a 섞기
idx = torch.randperm(100) # 0 ~ 99 섞어서 반환. -> index
idx
b = a[idx]

In [50]:
b

tensor([30, 27, 81, 84, 69,  9, 57, 22, 92, 62, 20, 23, 41, 48, 36,  0, 73, 55,
        17, 76, 25, 32, 29, 86, 98, 89, 67, 95, 99, 51, 64, 42, 39, 50, 21, 38,
        82, 28,  2, 83, 70,  4, 61, 54, 80, 47, 77, 90, 14, 33, 94, 91,  3, 18,
        19, 85, 68, 93, 65, 37, 79, 35, 74, 63, 31, 46, 72, 58, 56, 52, 78, 97,
        59, 66, 44,  5,  8, 16, 87, 24, 34,  1, 49, 71, 12, 15, 11, 40, 43, 10,
        13,  7, 60, 26, 96, 88,  6, 45, 75, 53])

In [52]:
b.device

device(type='cpu')

## Tensor gpu/cpu 메모리로 옮기기

- pytorch는 데이터셋인 tensor를 cpu메모리와 gpu 메모리로 서로 옮길 수 있다.
    - 데이터에 대한 연산처리를 어디서 하느냐에 따라 메모리를 선택한다.
    - 장치는 문자열로 설정한다.
        - CPU 사용: "cpu"
        - nvida GPU: "cuda"
        - Apple m1: "mps"
            - pytorch 1.12 부터 지원
- 옮기기
    - tensor 생성시 `device` 파라미터를 이용해 설정
    - `tensor.to(device)`를 이용해 설정
- 현재 실행환경에서 어떤 장비를 사용할 수 있는지 확인
    - nvidia gpu 사용가능확인
        - `torch.cuda.is_available()` - nvida gpu 사용가능 여부
        - `torch.backends.mps.is_available()` - M1 사용가능 여부

In [1]:
import torch
device = 'cuda' if torch.cuda.is_available() else 'cpu'

# device = 'mps' if torch.backends.mps.is_available() else 'cpu'
# device = 'cuda' if torch.cuda.is_available() else 'mps' if torch.backends.mps.is_available() else 'cpu'
device

'cuda'

In [3]:
t = torch.tensor([1, 2, 3], dtype=torch.float32, device=device) #생성할 때 device 지정.
print(t, t.device) # tensor가 어느 device에 있는지 확인

tensor([1., 2., 3.], device='cuda:0') cuda:0


In [4]:
t2 = t.to("cpu")  # 다른 device로 옮기기.
print(t2, t2.device)

tensor([1., 2., 3.]) cpu


In [5]:
t3 = t2.to("cuda")
print(t3, t3.device)

tensor([1., 2., 3.], device='cuda:0') cuda:0


## tensor를 상수로 변환
- torch.Tensor객체를 파이썬 상수로 변환.
- tensor객체.item()
    - Scalar(상수) 또는 원소가 하나인 tensor를 python 상수로 변환

In [1]:
import torch

In [2]:
print(torch.tensor(30))  # 0차원 텐서

tensor(30)


In [3]:
a = torch.tensor(10)  # 상수(scalar) => 0차원 Tensor
print(a, a.ndim)
print(a.item())
print(type(a), type(a.item()))

tensor(10) 0
10
<class 'torch.Tensor'> <class 'int'>


In [4]:
b = torch.tensor([[20]]) # 원소가 하나인 N차원 배열
print(b, b.dim())
print(b.item()) #원소가 하나인 배열(텐서) 변환 가능

tensor([[20]]) 2
20


In [6]:
c = torch.tensor([1, 10, 100])
print(c)
# print(c.item()) #원소가 여러개일 경우 Exception발생

tensor([  1,  10, 100])


In [ ]:
d = torch.tensor([10], device='cuda')
print(d)
print(d.item())

## ndarray 호환

- ndarray를 tensor로 생성
    - **torch.tensor(ndarray)**
    - **torch.from_numpy(ndarray)**
- tensor를 ndarray로 변환
    - **tensor.numpy()**
    - tensor가 gpu에 있을 경우 cpu로 옮긴 뒤 변환해야 한다.

In [2]:
import numpy as np
import torch

In [10]:
# ndarray -> tensor
arr = np.arange(1,10)

torch.tensor(arr, dtype=torch.float32)
torch.from_numpy(arr)

tensor([1, 2, 3, 4, 5, 6, 7, 8, 9])

In [11]:
# tensor -> ndarray
t = torch.randn(3,3)
print(t)
t.to("cpu").numpy()

tensor([[ 0.3503,  0.0878, -1.1883],
        [ 1.8476, -0.1706,  0.8687],
        [-0.3012,  0.2289,  0.1179]])


array([[ 0.35027522,  0.08784185, -1.1882582 ],
       [ 1.8476384 , -0.17064188,  0.8687183 ],
       [-0.3012033 ,  0.2288952 ,  0.11789653]], dtype=float32)

In [2]:
t2 = torch.randn(2,2, device="cuda")
t2

tensor([[-0.5223,  0.5846],
        [ 0.7011, -0.4219]], device='cuda:0')

In [6]:
# t2.numpy()
# t2.to("cpu").numpy()
t2.cpu().numpy()

array([[-0.5223277 ,  0.5846047 ],
       [ 0.70112723, -0.42187232]], dtype=float32)

# 원소 조회및 변경

## indexing/slicing

- 대부분 Numpy 와 동일
    - **slicing에서 step을 <u>음수로 지정할 수 없다.</u>**


In [3]:
t = torch.randint(-10, 10, (100, ))
t

tensor([ -6,  -5,  -5,   3,  -8,  -5,  -2,  -6,  -6,  -5,   4,   2, -10, -10,
          5,   1,   4,   3,  -1,  -1,   6,  -7,   3,  -6,  -3,   8,  -2,  -7,
         -7,  -2,  -5,  -1,   6,   9,  -4,   8,   6,   3, -10,  -4,  -7,   6,
        -10,  -8,  -6,  -7,  -7,   5,  -6,  -9,  -8,   9,  -2,   2,  -1,   6,
          6,  -7,   3,  -9,   7,   8,   8,   0,  -9,  -8,   3,  -5,  -3,   4,
          9,  -2, -10, -10,  -3,  -5,  -4,  -2,   2,   7,   6,   2,   1,  -9,
         -2,   0,  -5,   7,   0,   8,   6,   0,  -2,  -2,  -1,  -3,   0,  -4,
         -2,   3])

In [5]:
t[0]
t[[1, 5, -1]] # 여러개 조회-> fancy indexing

tensor([-5, -5,  3])

In [12]:
t[:5]
t[10:15]
t[90:]
t[3:30:3]

####### step 음수 안된다. 그래서 reverse(역순 조회)가 안됨. reverse하려면 flip() 사용
# t[10:1:-2]
t[1:10:2].flip(dims=(0,)) #reverse

tensor([-5, -6, -5,  3, -5])

## boolean index
- 논리 연산은 &, |, ~ 을 사용한다.

In [14]:
# boolean index
t[t > 0]

tensor([3, 4, 2, 5, 1, 4, 3, 6, 3, 8, 6, 9, 8, 6, 3, 6, 5, 9, 2, 6, 6, 3, 7, 8,
        8, 3, 4, 9, 2, 7, 6, 2, 1, 7, 8, 6, 3])

In [15]:
t[(t > 0) & (t < 5)]

tensor([3, 4, 2, 1, 4, 3, 3, 3, 2, 3, 3, 4, 2, 2, 1, 3])

In [16]:
# boolean index 처리 함수.
t.masked_select(t > 0)

tensor([3, 4, 2, 5, 1, 4, 3, 6, 3, 8, 6, 9, 8, 6, 3, 6, 5, 9, 2, 6, 6, 3, 7, 8,
        8, 3, 4, 9, 2, 7, 6, 2, 1, 7, 8, 6, 3])

In [17]:
t

tensor([ -6,  -5,  -5,   3,  -8,  -5,  -2,  -6,  -6,  -5,   4,   2, -10, -10,
          5,   1,   4,   3,  -1,  -1,   6,  -7,   3,  -6,  -3,   8,  -2,  -7,
         -7,  -2,  -5,  -1,   6,   9,  -4,   8,   6,   3, -10,  -4,  -7,   6,
        -10,  -8,  -6,  -7,  -7,   5,  -6,  -9,  -8,   9,  -2,   2,  -1,   6,
          6,  -7,   3,  -9,   7,   8,   8,   0,  -9,  -8,   3,  -5,  -3,   4,
          9,  -2, -10, -10,  -3,  -5,  -4,  -2,   2,   7,   6,   2,   1,  -9,
         -2,   0,  -5,   7,   0,   8,   6,   0,  -2,  -2,  -1,  -3,   0,  -4,
         -2,   3])

In [18]:
# 변경
t[0] = 100
t

tensor([100,  -5,  -5,   3,  -8,  -5,  -2,  -6,  -6,  -5,   4,   2, -10, -10,
          5,   1,   4,   3,  -1,  -1,   6,  -7,   3,  -6,  -3,   8,  -2,  -7,
         -7,  -2,  -5,  -1,   6,   9,  -4,   8,   6,   3, -10,  -4,  -7,   6,
        -10,  -8,  -6,  -7,  -7,   5,  -6,  -9,  -8,   9,  -2,   2,  -1,   6,
          6,  -7,   3,  -9,   7,   8,   8,   0,  -9,  -8,   3,  -5,  -3,   4,
          9,  -2, -10, -10,  -3,  -5,  -4,  -2,   2,   7,   6,   2,   1,  -9,
         -2,   0,  -5,   7,   0,   8,   6,   0,  -2,  -2,  -1,  -3,   0,  -4,
         -2,   3])

In [19]:
t = torch.arange(1, 10).reshape(3,3)
t

tensor([[1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]])

In [22]:
# t[ 0축, 1축 ]
t[1, 2].item()
t[0, 2].item()

3

In [23]:
t[[1,0], [2,2]]  #(1,2), (0,2)
# t[[첫번째값, 두번째값]-0축, [첫번째값, 두번째값]-1축]

tensor([6, 3])

# Reshape

## shape 변경
- tensor객체.reshape(\*shape) / view(\*shape) 이용
    - 변환 후 값을 변경하면 원본 배열의 값도 같이 바뀐다.
 > tensor.clone(): tensor를 복제한다.

In [24]:
a=torch.rand(12)  # (12, )
a2 = a.reshape(3,4)
a3 = a.reshape((3,2,2))
a4 = a.reshape((3,2,-1))  #한 개 axis는 -1로 설정가능하고 그럼 계산해서 알아서 설정해 준다.
print(a.shape, a2.size(), a3.shape, a4.shape)


torch.Size([12]) torch.Size([3, 4]) torch.Size([3, 2, 2]) torch.Size([3, 2, 2])


In [25]:
a5 = a.view(3,4)
a6 = a.view((3,2,2))
a7 = a.view((3,2,-1))  #한개 axis는 -1로 설정가능
print(a.shape, a5.size(), a6.shape, a7.shape)

torch.Size([12]) torch.Size([3, 4]) torch.Size([3, 2, 2]) torch.Size([3, 2, 2])


In [26]:
# a5[0, 0] = 12.1
a2[0, 1] = 15.1

print(a2)

tensor([[ 0.5366, 15.1000,  0.9277,  0.5999],
        [ 0.7028,  0.5628,  0.3314,  0.5761],
        [ 0.8905,  0.9121,  0.1569,  0.6140]])


In [28]:
a

tensor([ 0.5366, 15.1000,  0.9277,  0.5999,  0.7028,  0.5628,  0.3314,  0.5761,
         0.8905,  0.9121,  0.1569,  0.6140])

In [29]:
# tensor복사: clone() 메소드
r = a.reshape(3, 4).clone()
r[0,0] = 100.1
print(a)

tensor([ 0.5366, 15.1000,  0.9277,  0.5999,  0.7028,  0.5628,  0.3314,  0.5761,
         0.8905,  0.9121,  0.1569,  0.6140])


In [30]:
r

tensor([[100.1000,  15.1000,   0.9277,   0.5999],
        [  0.7028,   0.5628,   0.3314,   0.5761],
        [  0.8905,   0.9121,   0.1569,   0.6140]])

## dummy 축 늘리기

- None을 이용 (numpy의 newaxis 대신 None을 사용한다.)
- unsqueeze(dim=축번호)

In [38]:
import torch
a = torch.tensor([[10,20],[10,20]])
print(a.shape)

a1, a2 = a[None, :], a.unsqueeze(dim=0) # dim=정수만가능
print(a1.shape, a2.shape)

a3, a4 = a[:, :, None], a.unsqueeze(dim=-1)   # a[:, None] # 0축 :, 1축 None, 2번축 지정안함 -> 원래대로 그래서 (2, 1, 2) 가됨.
print(a3.shape, a4.shape)

a5, a6 = a3[:,None,:,:, None],  a3.unsqueeze(dim=1)  # unsqueeze(): dim=정수만가능 (한번에 dummy축은 하나만 추가가능.)
print(a5.shape, a6.shape)

torch.Size([2, 2])
torch.Size([1, 2, 2]) torch.Size([1, 2, 2])
torch.Size([2, 2, 1]) torch.Size([2, 2, 1])
torch.Size([2, 1, 2, 1, 1]) torch.Size([2, 1, 2, 1])


## dummy 축 제거
- squeeze(\[dim=축번호\]) 이용

In [42]:
t = torch.rand(3, 1, 4, 1, 5, 1)
print(t.shape)

r1 = t.squeeze()  #축을 명시 하지 않으면 모두 제거
print(r1.shape)

r2 = t.squeeze(dim=1) # 특정 axis 제거
print(r2.shape)

r3 = t.squeeze(dim=[1,3]) # 여러 axis의 dummy 축 제거
print(r3.shape)

torch.Size([3, 1, 4, 1, 5, 1])
torch.Size([3, 4, 5])
torch.Size([3, 4, 1, 5, 1])
torch.Size([3, 4, 5, 1])


# tensor 합치기
torch.cat([tensorA, tensorB, ...], dim=0) : 기준축 default 는 0
- 합치는 기준축 이외 축의 size가 동일해야하 합칠 수있다.

In [44]:
a = torch.arange(10).reshape(2,5)
b = torch.arange(10,20).reshape(2,5)
c = torch.arange(20,30).reshape(2,5)
d = torch.arange(10,19).reshape(3,3)

In [45]:
torch.cat([a, b], dim=0)

tensor([[ 0,  1,  2,  3,  4],
        [ 5,  6,  7,  8,  9],
        [10, 11, 12, 13, 14],
        [15, 16, 17, 18, 19]])

In [46]:
torch.cat([a, b, c], dim=0)

tensor([[ 0,  1,  2,  3,  4],
        [ 5,  6,  7,  8,  9],
        [10, 11, 12, 13, 14],
        [15, 16, 17, 18, 19],
        [20, 21, 22, 23, 24],
        [25, 26, 27, 28, 29]])

In [47]:
torch.cat([a, b], axis=1) # dim 대신 axis사용가능

tensor([[ 0,  1,  2,  3,  4, 10, 11, 12, 13, 14],
        [ 5,  6,  7,  8,  9, 15, 16, 17, 18, 19]])

In [48]:
torch.cat([a, b, c], axis=1)

tensor([[ 0,  1,  2,  3,  4, 10, 11, 12, 13, 14, 20, 21, 22, 23, 24],
        [ 5,  6,  7,  8,  9, 15, 16, 17, 18, 19, 25, 26, 27, 28, 29]])

In [49]:
torch.cat([a, b], axis=-1)  # -1: 마지막 axis

tensor([[ 0,  1,  2,  3,  4, 10, 11, 12, 13, 14],
        [ 5,  6,  7,  8,  9, 15, 16, 17, 18, 19]])

In [52]:
a.shape, d.shape
# 합치는 기준축 이외의 축 size는 같아야 한다.
torch.cat([a, d])  #dim=0, Error  1축 size가 달라서 Error

RuntimeError: Sizes of tensors must match except in dimension 0. Expected size 5 but got size 3 for tensor number 1 in the list.

# 값의 위치(index) 변경
- tensor 원소의 index의 위치를 바꾼다.
- `tensor.transpose(axis1, axis2)`
    - 두 축의 자리만 변경 할 수 있다.
- `tensor.permute(axis1, axis2, axis3, ..)`
    - 두 개 이상의 축 자리를 변경한다.

In [57]:
X = torch.arange(24).reshape(2, 3, 4)
print(X.shape)

y = X.transpose(1, 2)
print(y.shape)

z = X.permute(2, 0, 1)  # 2->0, 0->1, 1->2
print(z.shape)

torch.Size([2, 3, 4])
torch.Size([2, 4, 3])
torch.Size([4, 2, 3])


In [55]:
X # 2->0, 0->1, 1->2

tensor([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]])

In [58]:
z

tensor([[[ 0,  4,  8],
         [12, 16, 20]],

        [[ 1,  5,  9],
         [13, 17, 21]],

        [[ 2,  6, 10],
         [14, 18, 22]],

        [[ 3,  7, 11],
         [15, 19, 23]]])

In [56]:
y

tensor([[[ 0,  4,  8],
         [ 1,  5,  9],
         [ 2,  6, 10],
         [ 3,  7, 11]],

        [[12, 16, 20],
         [13, 17, 21],
         [14, 18, 22],
         [15, 19, 23]]])

# tensor 연산 및 주요 함수

## element-wise 연산
- tensor와 상수 연산시, tensor와 tensor간 연산시 원소별로 처리한다.
- 행렬곱 연산을 제외하고 tensor간 연산시 피연산자 tensor간에 shape이 같아야 한다.
    - shape이 다를 경우 조건이 맞으면 broadcasting을 한 뒤에 연산한다. (size가 다른 축의 경우 한개의 피연산자 size가 1일 경우 복사하여 shape을 맞춘다.)
    

In [59]:
import torch

a = torch.arange(10).reshape(2,5)
b = torch.arange(10,20).reshape(2,5)
c = torch.arange(50, 55)

print(a)
print(b)
print(c)

tensor([[0, 1, 2, 3, 4],
        [5, 6, 7, 8, 9]])
tensor([[10, 11, 12, 13, 14],
        [15, 16, 17, 18, 19]])
tensor([50, 51, 52, 53, 54])


In [60]:
print(a + 100)
print(a - 100)
print(a < 5)

tensor([[100, 101, 102, 103, 104],
        [105, 106, 107, 108, 109]])
tensor([[-100,  -99,  -98,  -97,  -96],
        [ -95,  -94,  -93,  -92,  -91]])
tensor([[ True,  True,  True,  True,  True],
        [False, False, False, False, False]])


In [61]:
print(a + b)
print(a == b)

tensor([[10, 12, 14, 16, 18],
        [20, 22, 24, 26, 28]])
tensor([[False, False, False, False, False],
        [False, False, False, False, False]])


In [62]:
a.size(), c.size()

(torch.Size([2, 5]), torch.Size([5]))

In [64]:
# broadcasting
print(a + c)

tensor([[50, 52, 54, 56, 58],
        [55, 57, 59, 61, 63]])


## 주요 연산함수

### 주요 상수
#### 자연상수 E

In [67]:
torch.e, np.e, torch.pi, np.pi

(2.718281828459045, 2.718281828459045, 3.141592653589793, 3.141592653589793)

#### torch.nan, torch.inf
- nan: Not a Number, 주로 결측치를 표현한다.
- inf: infinit 무한.
    - torch.inf: 양의 무한
    - -torch.inf: 음의 무한
- torch.isnan(tensor)
    - 원소별 결측치 확인
- torch.isinf(tensor)    
    - 원소별 inf 확인

In [68]:
print(torch.inf > 10000000000000000, torch.inf < 10000000000)
print(-torch.inf < -1000000000000000000000, -torch.inf > 10)


True False
True False


In [69]:
print(torch.log(torch.tensor(-1))) # nan (계산결과가 없으므로-없는값-nan 반환)
print(torch.isnan(torch.tensor([1,2,torch.nan,3,4])))  # nan 여부 확인
print(torch.isinf(torch.tensor([1,2,3,4,torch.inf])))  # inf 여부 확인

tensor(nan)
tensor([False, False,  True, False, False])
tensor([False, False, False, False,  True])


## 주요 연산 함수

In [70]:
x=torch.arange(-4, 5).reshape(3,3)
print(x)
print(torch.abs(x)) # 절대값
print(torch.sqrt(torch.abs(x))) #  제곱근
print(torch.exp(x))  # torch.e**x
print(torch.log(torch.abs(x)))
print(torch.log(torch.exp(torch.tensor(1))))  # torch.log() 밑이 e인 로그계산
print(torch.log10(torch.tensor(10)))         # torch.log10() 밑이 10인 로그계산
print(torch.log2(torch.tensor(2)))           # torch.log2() 밑이 2인 로그계산

tensor([[-4, -3, -2],
        [-1,  0,  1],
        [ 2,  3,  4]])
tensor([[4, 3, 2],
        [1, 0, 1],
        [2, 3, 4]])
tensor([[2.0000, 1.7321, 1.4142],
        [1.0000, 0.0000, 1.0000],
        [1.4142, 1.7321, 2.0000]])
tensor([[1.8316e-02, 4.9787e-02, 1.3534e-01],
        [3.6788e-01, 1.0000e+00, 2.7183e+00],
        [7.3891e+00, 2.0086e+01, 5.4598e+01]])
tensor([[1.3863, 1.0986, 0.6931],
        [0.0000,   -inf, 0.0000],
        [0.6931, 1.0986, 1.3863]])
tensor(1.)
tensor(1.)
tensor(1.)


In [71]:
y = x + torch.randn((3,3))
print(y)
print(torch.round(y)) # 반올림
print(torch.round(y, decimals=2)) # 소수점 둘째자리 이하에서 반올림
print(torch.floor(y)) # 내림
print(torch.ceil(y)) # 올림

tensor([[-7.3544, -3.9408, -1.9547],
        [ 0.1408, -0.1094,  1.2863],
        [-0.2881,  3.3387,  4.5116]])
tensor([[-7., -4., -2.],
        [ 0., -0.,  1.],
        [-0.,  3.,  5.]])
tensor([[-7.3500, -3.9400, -1.9500],
        [ 0.1400, -0.1100,  1.2900],
        [-0.2900,  3.3400,  4.5100]])
tensor([[-8., -4., -2.],
        [ 0., -1.,  1.],
        [-1.,  3.,  4.]])
tensor([[-7., -3., -1.],
        [ 1., -0.,  2.],
        [-0.,  4.,  5.]])


### 행렬곱
- `@` 연산자 또는 `torch.matmul(tensor1, tensor2)` 함수 이용

In [ ]:
import torch

In [72]:
x = torch.FloatTensor([[1, 2],[3, 4],[5, 6]])

y = torch.FloatTensor([[1, 2],[1, 2],])
x.size(), y.shape

(torch.Size([3, 2]), torch.Size([2, 2]))

In [73]:
z1 = x @ y
z2 = torch.matmul(x, y)
print(z1.shape, z2.shape)
print(z1)
print(z2)

torch.Size([3, 2]) torch.Size([3, 2])
tensor([[ 3.,  6.],
        [ 7., 14.],
        [11., 22.]])
tensor([[ 3.,  6.],
        [ 7., 14.],
        [11., 22.]])


In [74]:
# Batch 행렬곱(Batch matrix muliplication) - bmm()
# 피연산자로 두개의 3차원 tensor를 axis (1, 2)를 기준으로 행렬곱을 처리한다.
import torch
x = torch.FloatTensor(3,4,2)
y = torch.FloatTensor(3,2,5)
z = torch.bmm(x, y)
z.shape


torch.Size([3, 4, 5])

## 기술통계함수

In [75]:
X=torch.randn(3,4)
print(X)

tensor([[ 1.8222,  1.2650, -0.0240,  1.5413],
        [ 1.1235, -1.8762,  0.4169,  1.1345],
        [-0.2879,  1.5555, -1.3809, -0.7877]])


In [79]:
print(torch.sum(X))  # default: 전체기준으로 계산. dim=None
print(torch.sum(X, dim=0)) #dim/axis 지정: 지정한 axis의 index가 다른 값끼리 계산.
print(torch.sum(X, dim=0, keepdims=True)) # 기술통계함수들을 실행하면 차원이 줄어든다. keepdims=True로 하면 차원유지

tensor(4.5022)
tensor([ 2.6578,  0.9443, -0.9881,  1.8882])
tensor([[ 2.6578,  0.9443, -0.9881,  1.8882]])


In [80]:
print(torch.mean(X))
print(torch.mean(X, dim=0))
print(torch.mean(X, dim=0, keepdims=True))

tensor(0.3752)
tensor([ 0.8859,  0.3148, -0.3294,  0.6294])
tensor([[ 0.8859,  0.3148, -0.3294,  0.6294]])


In [81]:
print(torch.std(X)) # standard deviation 표준 편차
print(torch.var(X)) # variance
print(torch.var(X, dim=0))

tensor(1.2388)
tensor(1.5347)
tensor([1.1555, 3.6214, 0.8780, 1.5475])


In [82]:
# tensor.메소드()
print(X.sum(dim=1, keepdims=True))
print(X.mean(dim=1, keepdims=True))
print(X.std())

tensor([[ 4.6046],
        [ 0.7987],
        [-0.9011]])
tensor([[ 1.1511],
        [ 0.1997],
        [-0.2253]])
tensor(1.2388)


In [85]:
X.shape  # 3 X 4 (2차원)
X

tensor([[ 1.8222,  1.2650, -0.0240,  1.5413],
        [ 1.1235, -1.8762,  0.4169,  1.1345],
        [-0.2879,  1.5555, -1.3809, -0.7877]])

In [90]:
print(torch.max(X)) # 전체 기준 max값 계산.
print(torch.max(X, dim=0))  # 축을 지정하면 return_types.max 타입객체로 반환. max값과 max값의 index를 묶어서 반환
print(torch.max(X, dim=1))
print(torch.max(X, dim=1).values, torch.max(X, dim=1).indices, sep=" || ")


print(torch.max(X, dim=0, keepdims=True))  #keepdims=True : 차원(rank)를 유지
print(torch.max(X, dim=1, keepdims=True))

tensor(1.8222)
torch.return_types.max(
values=tensor([1.8222, 1.5555, 0.4169, 1.5413]),
indices=tensor([0, 2, 1, 0]))
torch.return_types.max(
values=tensor([1.8222, 1.1345, 1.5555]),
indices=tensor([0, 3, 1]))
tensor([1.8222, 1.1345, 1.5555]) || tensor([0, 3, 1])
torch.return_types.max(
values=tensor([[1.8222, 1.5555, 0.4169, 1.5413]]),
indices=tensor([[0, 2, 1, 0]]))
torch.return_types.max(
values=tensor([[1.8222],
        [1.1345],
        [1.5555]]),
indices=tensor([[0],
        [3],
        [1]]))


In [ ]:
# torch.return_types.max(
# values=tensor([1.8222, 1.5555, 0.4169, 1.5413]),
# indices=tensor([0, 2, 1, 0]))

In [93]:
print(torch.min(X))
print(torch.min(X, dim=0)) # return_types.min 타입객체로 반환. min값과 min값의 index를 묶어서 반환
print(torch.min(X, dim=1))

tensor(-1.8762)
torch.return_types.min(
values=tensor([-0.2879, -1.8762, -1.3809, -0.7877]),
indices=tensor([2, 1, 2, 2]))
torch.return_types.min(
values=tensor([-0.0240, -1.8762, -1.3809]),
indices=tensor([2, 1, 2]))


In [95]:
print(torch.min(X, dim=1).values)
print(torch.min(X, dim=1).indices)

tensor([-0.0240, -1.8762, -1.3809])
tensor([2, 1, 2])


In [99]:
print(torch.argmax(X)) # 전체중 가장 큰값의  index -> flatten(1차원으로 변환) 한 뒤 index를 반환.
print(torch.argmax(X, dim=0)) # 각 열에서 가장 큰 애가 존재하는 인덱스
print(torch.argmax(X, dim=1)) # 각 행에서 가장 큰 애가 존재하는 인덱스

tensor(0)
tensor([0, 2, 1, 0])
tensor([0, 3, 1])


In [100]:
a = torch.max(X, dim=0)
print(a.values)
print(a.indices)

tensor([1.8222, 1.5555, 0.4169, 1.5413])
tensor([0, 2, 1, 0])


# autograd(자동미분)
- 자동 미분을 이용해 gradient(미분계수)를 계산하는 pytorch system.
- 딥러닝 모델에서 weight와 bias tensor들(Parameter)은 backpropagation(역전파)를 이용해 gradient를 구해서 loss가 줄어드는 방향으로 update를 하게된다.
- pytorch는 이런 미분 수행을 자동으로 처리해 준다.
    - gradient(기울기)를 구한다는 것은 미분을 한다는 것을 말한다.
- tensor가 미분 가능하려면(gradient 계산 대상 변수) `requires_grad=True` 로 설정되 있어야 한다. (default: False)

> - **미분**
>   -  (순간) 변화율을 계산한다.
>   - $$\frac{\partial y}{\partial x}$$
>       - x에 대한 y의 변화율. 즉 x가 변하면 y는 얼마나 변할까?

In [101]:
import torch

In [103]:
x = torch.tensor([1.0], requires_grad=True)
print(x.requires_grad)
# Gradient를 계산할 대상인지 여부.
## True: x를 이용한 계산을 할때 gradient함수(도함수)를 정의해서 저장한다.
# x.reqires_grad = True

True


In [105]:
# Tensor의 속성
## data - Tenor가 가지는 값들을 저장하는 변수(속성).
print(x.data) # tensor([1.])
## grad - Gradient값을 저장하는 변수(속성)
print(x.grad)

tensor([1.])
None


In [107]:
y = x ** 2
y # tensor([1.], grad_fn=<PowBackward0>)

tensor([1.], grad_fn=<PowBackward0>)

In [108]:
y.grad_fn # gradient계산함수(도함수)가 저장된 변수(속성)
# dy / dx (x(requires_grad=True)에 대한 y의 변화율)

In [109]:
# gradient계산
y.backward() # dy/dx 계산 -> 그 결과를 x.grad에 저장.

In [111]:
x.data, x.grad, y.grad_fn
# (tensor([1.]), tensor([2.]), <PowBackward0 at 0x7d3f1e8e69b0>)

(tensor([1.]), tensor([2.]), <PowBackward0 at 0x7d3f1e8e69b0>)

In [119]:
a = torch.tensor([2.0], requires_grad=True)
b = torch.tensor([3.2]) # , requires_grad=True) # r_grad=False
y = a - b + 3  # grad_fn 정의
y.backward()   # grad를 계산 - a에 대해서만 계산.

In [120]:
y.grad_fn, a.data, a.grad

(<AddBackward0 at 0x7d3f1e8e7c40>, tensor([2.]), tensor([1.]))

In [121]:
print(b.grad)

tensor([-1.])


In [150]:
# 간단한 경사하강법 구현
input_data = torch.tensor([30.2]) # x
output_data = torch.tensor([10.5])# y

weight = torch.tensor([0.2489], requires_grad=True) # parameter

y_pred = weight * input_data  # 모델 예측값.

loss = (output_data - y_pred)**2 # 오차계산-MSE

In [151]:
print(y_pred.requires_grad)
y_pred, loss, y_pred.grad_fn, loss.grad_fn

True


(tensor([7.5168], grad_fn=<MulBackward0>),
 tensor([8.8996], grad_fn=<PowBackward0>),
 <PowBackward0 at 0x7d3f1e8e4d60>)

In [143]:
loss # 19.8916

tensor([19.8916], grad_fn=<PowBackward0>)

In [145]:
loss # 13.3051]

tensor([13.3051], grad_fn=<PowBackward0>)

In [152]:
loss #8.8996

tensor([8.8996], grad_fn=<PowBackward0>)

In [146]:
print(weight.grad)

None


In [147]:
loss.backward()

In [148]:
weight.grad

tensor([-220.3162])

In [149]:
new_weight = weight.data - weight.grad * 0.0001
new_weight

tensor([0.2489])

## torch.no_grad()
- no_grad()를 with 구문에서 연산을 할 경우 with block 내에서 requires_grad=True로 설정된 변수를 이용한 계산식이 있더라도 grad_fn을 생성하지 않는다. 즉 gradient를 계산하지 않는다.
- 딥러닝 모델을 평가 하거나 추론을할 때는 gradient를 계산할 필요가 없기 때문에 no_grad 구문을 사용한다.


In [160]:
x = torch.tensor([1.0], requires_grad=True)
# x.requires_grad = False
y = x**2
print("결과:", y.data)
print("grad_fn:", y.grad_fn)

결과: tensor([1.])
grad_fn: <PowBackward0 object at 0x7d3f1ea519f0>


In [161]:
print(x.requires_grad)
with torch.no_grad():
    y = x**2
    print("결과:", y.data)
    print("grad_fn:", y.grad_fn)

# True
# 결과: tensor([1.])
# grad_fn: None

True
결과: tensor([1.])
grad_fn: None


## gradient 값 초기화
- backward() 에서 계산되어 저장된 gradient값을 초기화한다.
- 반복 학습을 할 경우 gradient가 누적되는 문제가 발생한다. 그래서 한번 학습이 끝나면 다음 학습 전에 초기화하는 작업을 해야 한다.

In [169]:
x = torch.tensor([1.0], requires_grad=True)
print(x.grad)
y = x**2

y.backward()
print(x.grad)

None
tensor([2.])


In [174]:
# x를 이용한 새로운 계산
x.grad = None  # 초기화.
# grad값을 초기화 하지 않으면 이전 계산된 grad값이 남아있게 된다.
# 그리고 새로운 grad값을 구하게 되면 기존 값에 더한다.
z = x ** 3
z.backward()

In [175]:
print(x.grad) # tensor([5.])

tensor([3.])
